operator/pkg/helmreconciler/rendering.go

```diff
@@ -174,20 +197,87 @@ func unmarshalAndValidateIOPSpec(iopsYAML string) (*v1alpha1.IstioOperatorSpec,
 func (h *HelmReconciler) ProcessManifest(manifest manifest.Manifest) (int, error) {
 	var errs []error
-	log.Infof("Processing resources from manifest: %s", manifest.Name)
-	objects, err := object.ParseK8sObjectsFromYAMLManifest(manifest.Content)
+	crName := h.instance.Name + "-" + manifest.Name
+	log.Infof("Processing resources from manifest: %s for CR %s", manifest.Name, crName)
+	allObjects, err := object.ParseK8sObjectsFromYAMLManifest(manifest.Content)
 	if err != nil {
 		return 0, err
 	}
-	for _, obj := range objects {
+
+	objectCachesMu.Lock()
+
+	// Create and/or get the cache corresponding to the CR crName we're processing. Per crName partitioning is required to
+	// prune the cache to remove any objects not in the manifest generated for a given CR.
+	if objectCaches[crName] == nil {
+		objectCaches[crName] = &ObjectCache{
+			cache: make(map[string]*object.K8sObject),
+			mu:    &sync.RWMutex{},
+		}
+	}
+	objectCache := objectCaches[crName]
+
+	objectCachesMu.Unlock()
+
+	// Ensure that for a given CR crName only one control loop uses the per-crName cache at any time.
+	objectCache.mu.Lock()
+	defer objectCache.mu.Unlock()
+
+	// No further locking required beyond this point, since we have a ptr to a cache corresponding to a CR crName and no
+	// other controller is allowed to work on at the same time.
+
+	var changedObjects object.K8sObjects
+	var changedObjectKeys []string
+	allObjectsMap := make(map[string]bool)
+
+	// Check which objects in the manifest have changed from those in the cache.
+	for _, obj := range allObjects {
+		oh := obj.Hash()
+		allObjectsMap[oh] = true
+		if co, ok := objectCache.cache[oh]; ok && obj.Equal(co) {
+			// Object is in the cache and unchanged.
+			log.Infof("Object %s is unchanged, skip update.", oh)
+			continue
+		}
+		changedObjects = append(changedObjects, obj)
+		changedObjectKeys = append(changedObjectKeys, oh)
+	}
+
+	if len(changedObjectKeys) > 0 {
+		log.Infof("Changed object list: \n - %s", strings.Join(changedObjectKeys, "\n - "))
+	} else {
+		log.Infof("No objects changed for this component.")
+	}
+
+	// For each changed object, write it to the API server.
+	for _, obj := range changedObjects {
 		err = h.ProcessObject(manifest.Name, obj.UnstructuredObject())
 		if err != nil {
+			log.Error(err.Error())
 			errs = append(errs, err)
+			continue
+		}
+		log.Infof("Adding object %s to cache.", obj.Hash())
+		// Update the cache with the latest object.
+		objectCache.cache[obj.Hash()] = obj
+	}
+
+	// Prune anything not in the manifest out of the cache.
+	var removeKeys []string
+	for k := range objectCache.cache {
+		if !allObjectsMap[k] {
+			removeKeys = append(removeKeys, k)
 		}
 	}
-	return len(objects), utilerrors.NewAggregate(errs)
+	for _, k := range removeKeys {
+		log.Infof("Pruning object %s from cache.", k)
+		delete(objectCache.cache, k)
+	}
+
+	return len(changedObjects), utilerrors.NewAggregate(errs)
```

Comment:
This is problematic. If no objects were changed, because they are already in the right state, we don't report the component's status (because the stub health logic is just based on this number being >0)